daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala

```diff
@@ -160,7 +162,15 @@ abstract class ModelGroup(index: Int)
       groupMembers.forall { _.hasStaticallyRequiredOccurrencesInDataRepresentation }
   }
 
-  def groupMembers: Seq[Term]
+  final def turnOnSharing = true
+
+  final def groupMembers: Seq[Term] =
+    if (turnOnSharing && this.isSharingATerm)
+      sharedModelGroup.groupMembers1
+    else
+      groupMembers1
+
```

Comment:
The above is 1/2 the magic. If turnOnSharing, and this term has a shared proto-term, then use that to get the group members. This eliminates all the combinatorial explosion of creating DSOM objects recursively.